use builtin Fractions to help with Rational and igcd

smichr opened the pull request:
closes #11081 
closes #11089
- `igcd` now uses fractions.gcd to do the computation; caching is still in place. Also, a check that _all_ args be ints will catch cases where a gcd=1 might occur before bad input is encountered.
- the `Fraction` object is used to do handle some parsing of `Rational` input and to compute the limited denominator. I wonder if it might be better to use `Fraction` as the underlying representation and then update methods like `Rational.p` and `Rational.args` to return the correct args of the `Fraction`. Any thoughts would be welcome.
- [x] clean up history

smichr commented:
ping @isuruf -- please take a look at the mods here in contrast to #11089

isuruf commented on sympy/core/numbers.py:
```diff
@@ -155,24 +155,17 @@ def igcd(*args):
-
-            _gcdcache[(a, b)] = a
-        if a == 1 or b == 1:
+        a = _gcdcache.setdefault((a, b), gcd(a, b))
```
`gcd(a, b)` is computed each time `igcd(a, b)` is called.
```
try:
    a = _gcdcache[(a, b)]
except KeyError:
    a = _gcdcache[(a, b)] = gcd(a, b)
```

isuruf commented:
> I wonder if it might be better to use Fraction as the underlying representation and then update methods like Rational.p and Rational.args to return the correct args of the Fraction

Implementation of Fractions doesn't do the `gcd` hint trick introduced in #11019, therefore will be slower to use. Also, arithmetic operators would use `gcd` instead of `igcd` and therefore caching will not be used in that case.

isuruf commented on sympy/core/numbers.py:
```diff
@@ -155,24 +155,17 @@ def igcd(*args):
     5
 
     """
+    from fractions import gcd
```
You can also import the gcd function from `gmpy` if `HAS_GMPY` is true

smichr committed:
use Fraction for Rational handling; gcd for igcd

smichr commented:
ping @isuruf - ok, have a look now. I benchmark with

```
for I in range(10**4):
 for J in range(10**4):
   g = igcd(I,j)
```

I think it is now more similar to the original `igcd` now.

isuruf commented on sympy/core/numbers.py:
```diff
@@ -155,26 +155,32 @@ def igcd(*args):
+        b = args.pop()
+        if b == 0:
+            continue
+        b = as_int(b)
```
Why not have abs here?

smichr replied:
Although the sign change is handled below, handling it here would keep the cache from including negatives. I'll change it.

isuruf commented on sympy/core/numbers.py:
```diff
@@ -155,26 +155,32 @@ def igcd(*args):
+            if a == 1:
+                test_others = [as_int(a) for a in args]
+                return 1
+            _gcdcache[t] = a
```
This should come before line 179

smichr replied:
I was thinking not to cache the trivial arg but `a` is now the result. You're right. I'll move it.

smichr committed:
revamp igcd

modifications were tested with the following code to see that
performance was not degraded:

def go(N, number=3):
 from sympy.core.numbers import _gcdcache as G
 from time import time
 from random import randint as r
 G.clear()
 times = []
 for do in range(number):
  t = time()
  for i in range(N*N):
   igcd(r(0,N//2), r(0,N//2))
  times.append( time()-t)
 return min(times)

>>> go(1000)

Initial values were about 15 and are now about 10.

smichr commented:
I feel like the rewritten `igcd` is sound and ready for review, @isuruf and @jksuom . The code I used for testing is in the commit message.

jksuom commented on sympy/core/numbers.py:
```diff
@@ -155,25 +155,30 @@ def igcd(*args):
+            try:
+                a = _gcdcache[(a, b)]
+            except KeyError:
+                b = as_int(b)
```
Should this be done before looking into the cache?

smichr replied:
My thought was this: S(1) in {1:2} is True so the type doesn't matter. The usual use case is to use `igcd` on int/Integers so we can get the answer before doing a function call to `as_int`.

jksuom commented on sympy/core/numbers.py:
```diff
@@ -155,25 +155,30 @@ def igcd(*args):
-            _gcdcache[(a, b)] = a
-        if a == 1 or b == 1:
-            return 1
+                _gcdcache[t] = _gcdcache[t[1], t[0]] = a
```
Should `t[1], t[0]` be a tuple?

smichr replied:
It behaves as a tuple:

```
>>> d={};d[1,2]=3;d
{(1, 2): 3}
```

jksuom replied:
I thought so. (It was only that there are parentheses above on line 185.)

isuruf replied:
Instead of `_gcdcache[t] = _gcdcache[t[1], t[0]] = a`, you can simply do `_gcdcache[t] = a`

smichr replied:
By this line I am entering (a,b) and (b,a) into the cache.

isuruf replied:
+1

jksuom commented:
This looks good to me. +1

smichr committed:
enforce 2 args for igcd/ilcm; add quick exit for igcd(...,1,...)
